### PR TITLE
Fold refactor

### PR DESCRIPTION
**Pull request by ptpham**
Probably will require more changes in the future ...
If this is approved I will do another diff that will remove the old API and replace in all referenced locations.

### CODE REVIEW

**Comment by ptpham** [comment]
I'm also renaming HalfEdge to Halfedge because we use halfedge as a variable name everywhere. If HalfEdge were correct then we would have to change all the variable names to halfEdge. =P =P =P =P

**Comment by ptpham** [comment]
Hopefully I haven't destroyed the purity of your parsing code. O__O I'm a little skeptical about what I've replaced chai spies with. I'm having the Folds remember their arguments and then checking those things ... it's ... I don't know ... T__T

File: parser.js [hunk 1]
```diff
@@ -13,78 +14,92 @@ var geometry = require('./geometry.js');
...
+
+function splitEdges(paper, names, first, second) {
+  var halfedges = _.map([first, second], names.getEdge.bind(names));
+  var midpoints = _.map(halfedges, halfedge =>
+    geometry.midpoint(halfedge.src, halfedge.dst));
+  var directions = _.map(halfedges, (halfedge) =>
+    geometry.direction(halfedge.src, halfedge.dst));
+
+  var a = midpoints[0], c = midpoints[1];
+  var intersect = geometry.intersectLine(a, directions[0], c, directions[1]);
```
Comment (cixzhang) [inline parser.js]: splitEdges is used for edge to edge folds. These edges might not intersect.

File: test/fold.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
+
+
+var EPSILON = require('../constants').EPSILON;
+var Paper = require('../paper');
```
Comment (cixzhang) [inline test/fold.js]: Do we really need all of these for this test?

File: parser.js (hunk 1 above)
Comment (ptpham) [inline parser.js]: Hmm ... good point. I will need to think about this.

File: test/fold.js (hunk 2 above)
Comment (ptpham) [inline test/fold.js]: =P I'll get rid of the unnecessary ones.

File: geometry.js [hunk 3]
```diff
@@ -20,6 +20,30 @@ function bisectAngle(vA, vB, vC) {
...
 
+function bisectQuad(a, b, c, d) {
+  if (distance(a, c) > distance(b, c)) {
+    var temp = a;
+    a = b;
+    b = temp;
+  }
+
+  if (distance(c, a) > distance(d, a)) {
+    var temp = c;
```
Comment (cixzhang) [inline geometry.js]: We're declaring this variable twice.

File: parser.js [hunk 4]
```diff
@@ -13,78 +14,118 @@ var geometry = require('./geometry.js');
...
-      parsed.halfedges = sheet.seamFlood(halfedges);
-      parsed.vertex = line.vertex;
-      parsed.direction = line.direction;
-      return parsed;
-    }
-  }
+
+Parser.parse = parse;
+Parser.prototype.parse = function(string) {
+  var parsed = parse(string), fold = this.fold;
```
Comment (cixzhang) [inline parser.js]: Not using fold variable?

**Comment by cixzhang** [comment]
I think this looks good! :D

**Pull request merged**